filters/encrypt/filter.go

```diff
@@ -0,0 +1,554 @@
+	case reflect.Ptr, reflect.Interface:
+		if payloadValue.IsNil() { // be sure it's not a nil interface
+			return e, nil
+		}
+		payloadValue = reflect.ValueOf(e.Payload).Elem()
+	}
+
+	pType := payloadValue.Type()
+	pKind := payloadValue.Kind()
+
+	taggedInterface, isTaggable := payloadValue.Interface().(Taggable)
+
+	// make a copy of the overrides before we begin processing this event, which
+	// will give us a consistent set of overrides for this event.
+	filterOverrides := ef.copyFilterOperationOverrides()
+
+	switch {
+	case pType == reflect.TypeOf("") || pType == reflect.TypeOf([]uint8{}):
+		if !payloadValue.CanSet() {
+			return nil, fmt.Errorf("%s: unable to redact string payload (not setable): %w", op, ErrInvalidParameter)
+		}
+		classificationTag := getClassificationFromTagString(string(SecretClassification), withFilterOperations(filterOverrides))
+		if err := ef.filterValue(ctx, payloadValue, classificationTag, opts...); err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+	case isTaggable:
+		if err := ef.filterTaggable(ctx, taggedInterface, opts...); err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+	case pKind == reflect.Slice:
+		switch {
+		// if the field is a slice of string or slice of []byte
+		case pType == reflect.TypeOf([]string{}) || pType == reflect.TypeOf([]*string{}) || pType == reflect.TypeOf([][]uint8{}):
+			classificationTag := getClassificationFromTagString(string(SecretClassification), withFilterOperations(filterOverrides))
+			if err := ef.filterSlice(ctx, classificationTag, payloadValue, opts...); err != nil {
+				return nil, fmt.Errorf("%s: %w", op, err)
+			}
+		// if the field is a slice of structs, recurse through them...
+		default:
+			for i := 0; i < payloadValue.Len(); i++ {
+				f := payloadValue.Index(i)
+				if f.Kind() == reflect.Ptr {
+					f = f.Elem()
+				}
+				if f.Kind() != reflect.Struct {
+					continue
+				}
+				if err := ef.filterField(ctx, f, filterOverrides, opts...); err != nil {
+					return nil, fmt.Errorf("%s: %w", op, err)
+				}
+			}
+		}
+	case pKind == reflect.Struct:
+		if err := ef.filterField(ctx, payloadValue, filterOverrides, opts...); err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+	}
+	return e, nil
+}
+
+func (ef *Filter) copyFilterOperationOverrides() map[DataClassification]FilterOperation {
+	if ef.FilterOperationOverrides == nil {
+		return nil
+	}
+	ef.l.RLock()
+	defer ef.l.RUnlock()
+	cp := map[DataClassification]FilterOperation{}
+	for k, v := range ef.FilterOperationOverrides {
+		cp[k] = v
+	}
+	return cp
+}
+
+// filterField will recursively iterate over all the fields for a struct value
+// and filter them based on their DataClassification
+func (ef *Filter) filterField(ctx context.Context, v reflect.Value, filterOverrides map[DataClassification]FilterOperation, opt ...Option) error {
+	const op = "event.(Filter).filterField"
+	// check for nil value (prevent panics)
+	if v == reflect.ValueOf(nil) {
```

Comment:
How come you used ValueOf(nil) here instead of IsNil()?

Reply:
Same question in various places below.

Reply:
you can't call IsNil() on a struct value (it will panic).  You can only use it if v is: chan, func, interface, map, pointer or slice

Reply:
Ah cool, thanks.